> > This will only work for the testcase in bin/test core
> 
> Which test do you mean?

The one which was failing in bin/test core, because that is the only case which passed after making this change.

```
 sympy\core\tests\test_expr.py:test_as_leading_term4 ______________
Traceback (most recent call last):
 sympy\core\tests\test_expr.py", line 485, in test_as_leading_term4
    assert r.as_leading_term(x).cancel() == n/2
AssertionError

here if we write .cancel() before .as_leading_term(x) then it will automatically work because it will then first simplify the expression and then pass it to the .as_leading_term(x) function
```

--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -861,7 +861,7 @@ class Add(Expr, AssocOp):
         return (self.func(*re_part), self.func(*im_part))
 
     def _eval_as_leading_term(self, x):
-        from sympy import expand_mul, factor_terms, Order
+        from sympy import expand_mul, Order
 
         old = self
 
@@ -874,28 +874,28 @@ class Add(Expr, AssocOp):
 
         leading_terms = [t.as_leading_term(x) for t in expr.args]
 
+        min, new_expr = Order(0), 0
+
         try:
-            min = Order(leading_terms[0], x).getn()
+            for term in leading_terms:
+                order = Order(term, x)
+                if order not in min:
+                    min = order
+                    new_expr = term
+                elif order == min:
+                    new_expr += term
+
         except Exception:
             compute = True
-        else:
-            new_expr = leading_terms[0]
 
-        if compute == False:
-            for term in leading_terms[1:]:
-                try:
-                    order = Order(term, x).getn()
-                except Exception:
-                    compute = True
-                    break
-                else:
-                    if order < min:
-                        min = order
-                        new_expr = term
-                    elif order == min:
-                        new_expr += term
+        if compute:
+            return expr
 
-        if compute or not new_expr:
+        new_expr=new_expr.together()
+        if new_expr.is_Add:
+            new_expr = new_expr.simplify()
+
+        if not new_expr:
             # simple leading term analysis gave us cancelled terms but we have to send
             # back a term, so compute the leading term (via series)
             return old.compute_leading_term(x)

--- a/sympy/core/add.py
+++ b/sympy/core/add.py
@@ -870,7 +870,6 @@ class Add(Expr, AssocOp):
             return expr.as_leading_term(x)
 
         infinite = [t for t in expr.args if t.is_infinite]
-        compute = False
 
         leading_terms = [t.as_leading_term(x) for t in expr.args]
 
@@ -879,16 +878,13 @@ class Add(Expr, AssocOp):
         try:
             for term in leading_terms:
                 order = Order(term, x)
-                if order not in min:
+                if not min or order not in min:
                     min = order
                     new_expr = term
                 elif order == min:
                     new_expr += term
 
-        except Exception:
-            compute = True
-
-        if compute:
+        except TypeError:
             return expr
 
         new_expr=new_expr.together()

--- a/sympy/core/function.py
+++ b/sympy/core/function.py
@@ -706,6 +706,9 @@ class Function(Application, Expr):
             if e == e1:
                 #for example when e = sin(x+1) or e = sin(cos(x))
                 #let's try the general algorithm
+                if len(e.args) == 1:
+                    # issue 14411
+                    e = e.func(e.args[0].cancel())
                 term = e.subs(x, S.Zero)
                 if term.is_finite is False or term is S.NaN:
                     raise PoleError("Cannot expand %s around 0" % (self))

--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -1444,11 +1444,13 @@ class Pow(Expr):
 
                 # express "rest" as: rest = 1 + k*x**l + ... + O(x**n)
                 rest = expand_mul((b - prefactor)/prefactor)
+                rest = rest.simplify()
 
                 if rest.is_Order:
                     return 1/prefactor + rest/prefactor + O(x**n, x)
 
                 k, l = rest.leadterm(x)
+
                 if l.is_Rational and l > 0:
                     pass
                 elif l.is_number and l > 0:

--- a/sympy/functions/elementary/exponential.py
+++ b/sympy/functions/elementary/exponential.py
@@ -884,7 +884,7 @@ class log(Function):
     def _eval_nseries(self, x, n, logx):
         # NOTE Please see the comment at the beginning of this file, labelled
         #      IMPORTANT.
-        from sympy import cancel, Order
+        from sympy import cancel, Order, logcombine
         if not logx:
             logx = log(x)
         if self.args[0] == x:
@@ -905,6 +905,8 @@ class log(Function):
             s = self.args[0].nseries(x, n=n, logx=logx)
         a, b = s.leadterm(x)
         p = cancel(s/(a*x**b) - 1)
+        if p.has(exp):
+            p = logcombine(p)
         g = None
         l = []
         for i in range(n + 2):

--- a/sympy/solvers/recurr.py
+++ b/sympy/solvers/recurr.py
@@ -813,10 +813,10 @@ def rsolve(f, y, init=None):
                     i = int(k.args[0])
                 else:
                     raise ValueError("Integer or term expected, got '%s'" % k)
-            try:
+
+            eq = solution.limit(n, i) - v
+            if eq.has(S.NaN):
                 eq = solution.limit(n, i) - v
-            except NotImplementedError:
-                eq = solution.subs(n, i) - v
             equations.append(eq)
 
         result = solve(equations, *symbols)

